### PR TITLE
test:add a testcase to verify connections of etcd

### PR DESCRIPTION
**Pull request by idbeta**
### What this PR does / why we need it:
when used APISIX with etcd v3,there will cause create too many connections of etcd, https://github.com/apache/apisix/issues/2295
so i add a testcase of pytest for verify it.

### Pre-submission checklist:

* [x] Did you explain what problem does this PR solve? Or what new features have been added?
* [] Have you added corresponding test cases?
* [] Have you modified the corresponding document?
* [x] Is this PR backward compatible?

### CODE REVIEW

**Comment by moonming** [comment]
please answer the question first:
**What this PR does / why we need it**

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -12,7 +12,7 @@ jobs:
       fail-fast: false
       matrix:
         platform: [ubuntu-18.04]
-        os_name: [linux_openresty, linux_tengine, linux_apisix_master_luarocks, linux_apisix_current_luarocks, linux_openresty_mtls]
+        os_name: [linux_openresty, linux_tengine, linux_apisix_master_luarocks, linux_apisix_current_luarocks, linux_openresty_mtls, linux_apisix_integrationtest]
```
Comment (moonming) [inline .github/workflows/build.yml:15]: `linux_apisix_integrationtest`  we can remove the `linux_apisix_` prefix.

Comment (moonming) [reply .github/workflows/build.yml:15]: how about `e2e_test`?

File: .travis/linux_apisix_integrationtest_runner.sh [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+
+export_or_prefix() {
+    export OPENRESTY_PREFIX="/usr/local/openresty-debug"
+}
+
+create_lua_deps() {
+    echo "Create lua deps cache"
+
+    make deps
+    luarocks install luacov-coveralls --tree=deps --local > build.log 2>&1 || (cat build.log && exit 1)
```
Comment (moonming) [inline .travis/linux_apisix_integrationtest_runner.sh:29]: should remove useless code.

File: .travis/linux_apisix_integrationtest_runner.sh [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+
+    sudo rm -rf build-cache/deps
+    sudo cp -r deps build-cache/
+    sudo cp rockspec/apisix-master-0.rockspec build-cache/
+}
+
+before_install() {
+    sudo cpanm --notest Test::Nginx >build.log 2>&1 || (cat build.log && exit 1)
```
Comment (moonming) [inline .travis/linux_apisix_integrationtest_runner.sh:37]: ditto

File: .travis/linux_apisix_integrationtest_runner.sh [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+    docker pull redis:3.0-alpine
+    docker run --rm -itd -p 6379:6379 --name apisix_redis redis:3.0-alpine
+    docker run --rm -itd -e HTTP_PORT=8888 -e HTTPS_PORT=9999 -p 8888:8888 -p 9999:9999 mendhak/http-https-echo
+    # Runs Keycloak version 10.0.2 with inbuilt policies for unit tests
+    docker run --rm -itd -e KEYCLOAK_USER=admin -e KEYCLOAK_PASSWORD=123456 -p 8090:8080 -p 8443:8443 sshniro/keycloak-apisix
```
Comment (moonming) [inline .travis/linux_apisix_integrationtest_runner.sh:42]: ditto.
please check this cript one by one line, only keep what you need

File: t/integrationtest/cases/test_base.py [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+#!/usr/bin/env python
+#-*- coding: utf-8 -*-
+import sys,os,time,requests,json,subprocess,signal,psutil,grequests
```
Comment (moonming) [inline t/integrationtest/cases/test_base.py:20]: not following PEP8.
please check python code style first.

File: t/integrationtest/cases/test_base.py [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+
+def killprocesstree(id):
+    for pid in psutil.pids():
+        if psutil.Process(int(pid)).ppid()==id:
+            psutil.Process(int(pid)).terminate()
+    psutil.Process(id).terminate()
+
+def getpidbyname():
```
Comment (moonming) [inline t/integrationtest/cases/test_base.py:28]: get_pid_by_name

File: t/integrationtest/cases/test_base.py [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+    name = "apisix"
+    cmd = "ps -ef | grep %s | grep nginx | grep -v grep | grep -v %s/t | awk '{print $2}'"%(name,name)
+    p = subprocess.Popen(cmd, stderr = subprocess.PIPE, stdout = subprocess.PIPE, shell = True)
+    p.wait()
+    return p.stdout.read().strip()
+
+def getworkerres(pid):
```
Comment (moonming) [inline t/integrationtest/cases/test_base.py:35]: ditto

File: t/integrationtest/cases/test_base.py [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+    headers = {"X-API-KEY": "edd1c9f034335f136f87ad84b625c8f1"}
+    casepath = cur_file_dir()
+    confpath = casepath + "/nginx.conf"
+    try:
+        os.makedirs("./cases/logs")
+    except Exception as e:
+        pass
+    p = subprocess.Popen(['openresty', '-p',casepath,'-c',confpath], stderr = subprocess.PIPE, stdout = subprocess.PIPE, shell = False)
+    p.wait()
+    nginx_pid = p.pid+1
```
Comment (moonming) [inline t/integrationtest/cases/test_base.py:68]: you can not get nginx pid by this method.
please get from `.pid` file.

File: t/integrationtest/cases/test_base.py [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+
+def teardown_module():
+    pass
+    # killprocesstree(nginx_pid)
+
+def test_basescenario01():
```
Comment (moonming) [inline t/integrationtest/cases/test_base.py:74]: bad name

File: t/integrationtest/cases/test_base.py [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+    print("APISIX's resource occupation(before test):")
+    getworkerres(apisixpid)
+    cfgdata = {
```
Comment (moonming) [inline t/integrationtest/cases/test_base.py:77]: what is `cfg`?

File: t/integrationtest/cases/test_base.py [hunk 11]
```diff
@@ -0,0 +1,106 @@
+#
```
Comment (moonming) [inline t/integrationtest/cases/test_base.py:1]: Is this file related to PR?

File: t/integrationtest/cases/test_etcdlinks.py [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+
+def cur_file_dir():
+    return os.path.split(os.path.realpath(__file__))[0]
+
+def setup_module():
+    global headers,nginx_pid,apisixhost,apisixpid,apisixpath
+    apisixpid = int(getpidbyname())
+    apisixpath = psutil.Process(apisixpid).cwd()
+    apisixhost = "http://127.0.0.1:9080"
+    headers = {"X-API-KEY": "edd1c9f034335f136f87ad84b625c8f1"}
```
Comment (moonming) [inline t/integrationtest/cases/test_etcdlinks.py:55]: are those codes all useless?

**Review by moonming — CHANGES_REQUESTED** [review]

**Comment by moonming** [comment]
@idbeta you submit many useless codes. please ONLY submit code relate with this PR.

**Comment by liuhengloveyou** [comment]
good for me. 45 commit histories can be rebased into 1.

File: .github/workflows/build.yml (hunk 1 above)
Comment (idbeta) [reply .github/workflows/build.yml:15]: that 's good idea! I will change it.

File: .travis/linux_apisix_integrationtest_runner.sh (hunk 2 above)
Comment (idbeta) [reply .travis/linux_apisix_integrationtest_runner.sh:29]: > should remove useless code.

Yes, here I directly reuse the previous shell script,there ara extra code that has not been removed.

File: t/integrationtest/cases/test_base.py (hunk 5 above)
Comment (idbeta) [reply t/integrationtest/cases/test_base.py:20]: > not following PEP8.
> please check python code style first.
the code here is really ugly,i will change it

File: t/integrationtest/cases/test_base.py (hunk 6 above)
Comment (idbeta) [reply t/integrationtest/cases/test_base.py:28]: > get_pid_by_name

Code style issues,got it.

File: t/integrationtest/cases/test_base.py (hunk 8 above)
Comment (idbeta) [reply t/integrationtest/cases/test_base.py:68]: > you can not get nginx pid by this method.
> please get from `.pid` file.

There is no need to get nginx_pid here,i will remove it.

File: t/integrationtest/cases/test_base.py (hunk 10 above)
Comment (idbeta) [reply t/integrationtest/cases/test_base.py:77]: > @idbeta you submit many useless codes. please ONLY submit code relate with this PR.

my fault,I should remove irrelevant code,actually,test_base.py has nothing to do with this PR

File: t/integrationtest/cases/test_etcdlinks.py [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+    casepath = cur_file_dir()
+
+def teardown_module():
+    pass
+
+def test_etcdlinks01():
+    time.sleep(30)
+    assert getetcdlinks() <= getworkernum(apisixpid)*7
```
Comment (membphis) [inline t/integrationtest/cases/test_etcdlinks.py:62]: The function name style is bad, please update it to `xxx_xxxx_xxxx' style, and fix similar points.
this name is hard for reading.

Comment (membphis) [reply t/integrationtest/cases/test_etcdlinks.py:62]: and please add a space between the different objects.

eg, `getworkernum(apisixpid)*7` -> `getworkernum(apisixpid) * 7`

**Review by membphis — CHANGES_REQUESTED** [review]

**Comment by membphis** [comment]
@idbeta https://github.com/apache/apisix/pull/2329/checks?check_run_id=1174617370#step:6:132

the CI is failed, please take a look at the output.

**Comment by idbeta** [comment]
> @idbeta https://github.com/apache/apisix/pull/2329/checks?check_run_id=1174617370#step:6:132
> 
> the CI is failed, please take a look at the output.

awkward,i forgot that,i fix it now

**Comment by idbeta** [comment]
This pr contains too much useless code, it should not be discussed anymore, close it
